tests/python_tests/helpers/profiler.py

```diff
@@ -0,0 +1,186 @@
+# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from dataclasses import dataclass
+from typing import Optional
+import re
+
+from ttexalens.tt_exalens_lib import read_words_from_device
+
+from helpers.test_config import generate_make_command, ProfilerBuild
+from helpers.utils import run_shell_command
+
+
+def _hash_profiler_message(s: str) -> int:
+    hash32 = 2166136261
+    for c in s.encode("ascii"):
+        hash32 ^= c
+        hash32 = (hash32 * 16777619) & 0xFFFFFFFF  # simulate 32-bit unsigned overflow
+    return (hash32 ^ (hash32 >> 16)) & 0xFFFF  # fold to 16 bits
+
+
+_PROFILER_PATTERN = re.compile(
+    r"'#pragma message: (?P<full_marker>LLK_PROFILER:(?P<file>[^:]+):(?P<line>\d+):(?P<marker>[^']+))'",
+)
+
+
+@dataclass
+class ProfilerFullMarker:
+    marker: str
+    file: str
+    line: int
+    id: int
+
+
+def _process_profiler_message(line: str):
+    # ex. '#pragma message: LLK_PROFILER:sources/example.cpp:1337:MARKER'
+    expr = _PROFILER_PATTERN.search(line)
+    if expr is None:
+        return None
+
+    groups = expr.groupdict()
+    return ProfilerFullMarker(
+        marker=groups["marker"],
+        file=groups["file"],
+        line=int(groups["line"]),
+        id=_hash_profiler_message(groups["full_marker"]),
+    )
+
+
+def build_perf_test(test_config):
+    make_cmd = generate_make_command(test_config, ProfilerBuild.Yes)
+    result = run_shell_command(f"cd .. && {make_cmd}")
+
+    lines = result.stderr.splitlines()
+    return {
+        message.id: message
+        for line in lines
+        if (message := _process_profiler_message(line))
+    }
+
+
+@dataclass
+class ProfilerTimestamp:
+    full_marker: ProfilerFullMarker
+    timestamp: int
+    data: Optional[int]
```

Comment:
You can default this to `None`, so you don't have to pass None when there's no data, i.e.
```suggestion
    data: Optional[int] = None
```